test/functional/test_framework/authproxy.py

```diff
@@ -79,6 +80,14 @@ def __init__(self, service_url, service_name=None, timeout=HTTP_TIMEOUT, connect
         self.timeout = timeout
         self._set_conn(connection)
 
+    def with_auth(self, user, password):
+        authpair = user + ':' + password
+        authpair.replace('\\', '\\\\')
+        new_url = '%s' % (self.__service_url,)
+        # Based on RFC 3986 (I can't believe there isn't a better way to do this!)
```

Comment:
Untested:
```py
u = urlparse(self.__service_url)
u = u._replace(netloc="{}:{}@{}".format(user, password, u.hostname))
return AuthServiceProxy(u.geturl(), ...)
```

Reply:
That looks worse (readability, undocumented/undefined behaviour, etc)

Reply:
It's documented in https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse

> As is the case with all named tuples, the subclass has a few additional methods and attributes that are particularly useful. One such method is _replace(). The _replace() method will return a new ParseResult object replacing specified fields with new values.

https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urllib.parse.SplitResult.geturl

Reply:
Hmm, okay...

Might still need a regex, though, since hostname/port are not necessarily obvious.

Consider `https://a:b@[ffff:aaaa]:1234`